reactor-core/src/main/java/reactor/core/publisher/FluxSwitchOnFirst.java

```diff
@@ -19,610 +19,1029 @@
+	}
+
+	static boolean hasOutboundSubscribed(long state) {
+		return (state & HAS_OUTBOUND_SUBSCRIBED_FLAG) == HAS_OUTBOUND_SUBSCRIBED_FLAG;
+	}
+
+	static boolean hasOutboundCancelled(long state) {
+		return (state & HAS_OUTBOUND_CANCELLED_FLAG) == HAS_OUTBOUND_CANCELLED_FLAG;
+	}
+
+	static boolean hasOutboundTerminated(long state) {
+		return (state & HAS_OUTBOUND_TERMINATED_FLAG) == HAS_OUTBOUND_TERMINATED_FLAG;
+	}
+
+	static abstract class AbstractSwitchOnFirstMain<T, R> extends Flux<T>
+			implements InnerOperator<T, R> {
+
+		final ControlSubscriber<? super R>                                     outer;
+		final BiFunction<Signal<? extends T>, Flux<T>, Publisher<? extends R>>
+		                                                                       transformer;
+
+		Subscription s;
+
+		boolean isInboundRequestedOnce;
+		T       firstValue;
+
+		Throwable throwable;
+		boolean   done;
+
+		CoreSubscriber<? super T> inner;
+
+		volatile     int                                                  state;
+		@SuppressWarnings("rawtypes")
+		static final AtomicIntegerFieldUpdater<AbstractSwitchOnFirstMain> STATE =
+				AtomicIntegerFieldUpdater.newUpdater(AbstractSwitchOnFirstMain.class, "state");
+
+		@SuppressWarnings("unchecked")
+		AbstractSwitchOnFirstMain(CoreSubscriber<? super R> outer,
+				BiFunction<Signal<? extends T>, Flux<T>, Publisher<? extends R>> transformer,
+				boolean cancelSourceOnComplete) {
+			this.outer = outer instanceof Fuseable.ConditionalSubscriber ?
+					new SwitchOnFirstConditionalControlSubscriber<>(this,
+							(Fuseable.ConditionalSubscriber<R>) outer,
+							cancelSourceOnComplete) :
+					new SwitchOnFirstControlSubscriber<>(this,
+							outer,
+							cancelSourceOnComplete);
+			this.transformer = transformer;
+		}
+
+		@Override
+		@Nullable
+		public Object scanUnsafe(Attr key) {
+			final boolean isCancelled = this.inner == Operators.EMPTY_SUBSCRIBER;
+
+			if (key == Attr.CANCELLED) {
+				return isCancelled && !this.done;
+			}
+			if (key == Attr.TERMINATED) {
+				return this.done || isCancelled;
+			}
+			if (key == Attr.RUN_STYLE) {
+				return Attr.RunStyle.SYNC;
+			}
+
+			return InnerOperator.super.scanUnsafe(key);
+		}
+
+		@Override
+		public CoreSubscriber<? super R> actual() {
+			return this.outer;
+		}
+
+		@Override
+		public void onSubscribe(Subscription s) {
+			if (Operators.validate(this.s, s)) {
+				this.s = s;
+				this.outer.sendSubscription();
+				if (!hasInboundCancelled(this.state)) {
+					s.request(1);
+				}
+			}
+		}
+
+		@Override
+		public void onNext(T t) {
+			if (this.done) {
+				Operators.onNextDropped(t, currentContext());
+				return;
+			}
+
+			final T f = this.firstValue;
+			if (f == null) {
+				this.firstValue = t;
+
+				final long previousState = markFirstValueReceived(this);
+				if (hasInboundCancelled(previousState)) {
+					Operators.onDiscard(t, this.outer.currentContext());
```

Comment:
is there really no situation where `this.firstValue` wouldn't get nulled out externally? ie. does it hurt to explicitly null it out here?

 I think we need to be mindful of object retention here, on top of discarding (even though the `state` ensures the field is never read again). wdyt?